File: src/main/kotlin/org/skialin/GraphiteBackendTexture.kt
package org.skialin

import org.skialin.impl.Managed
import org.skialin.impl.NativeLoader

class GraphiteBackendTexture internal constructor(
    ptr: Long,
) : Managed(ptr, GraphiteBackendTextureNative::nRelease) {
    val isValid: Boolean get() = GraphiteBackendTextureNative.nIsValid(nativePtr)

    companion object {
        /**
         * Wraps a caller-owned `id<MTLTexture>` ([texture] is the native pointer). Not retained:
         * the caller keeps it alive for as long as this is in use. Null off macOS.
         */
        fun makeMetal(
            width: Int,
            height: Int,
            texture: Long,
        ): GraphiteBackendTexture? {
            val ptr = GraphiteBackendTextureNative.nMakeMetal(width, height, texture)
            return if (ptr == 0L) null else GraphiteBackendTexture(ptr)
        }

        /** Wraps a caller-owned VkImage (not allocated or freed by Skia). */
        @Suppress("LongParameterList")
        fun makeVk(
            width: Int,
            height: Int,
            sampleCount: Int,
            mipmapped: Boolean,
            imageCreateFlags: Int,
            format: Int,
            imageTiling: Int,
            imageUsageFlags: Int,
            sharingMode: Int,
            aspectMask: Int,
            currentLayout: Int,
            queueFamilyIndex: Int,
            image: Long,
            allocMemory: Long = 0L,
            allocOffset: Long = 0L,
            allocSize: Long = 0L,
            allocFlags: Int = 0,
        ): GraphiteBackendTexture {
            val ptr =
                GraphiteBackendTextureNative.nMakeVk(
                    width,
                    height,
                    sampleCount,
                    mipmapped,
                    imageCreateFlags,
                    format,
                    imageTiling,
                    imageUsageFlags,
                    sharingMode,
                    aspectMask,
                    currentLayout,
                    queueFamilyIndex,
                    image,
                    allocMemory,
                    allocOffset,
                    allocSize,
                    allocFlags,
                )
            return GraphiteBackendTexture(ptr)
        }
    }
}

private object GraphiteBackendTextureNative {
    init {
        NativeLoader.ensureLoaded()
    }

    @Suppress("LongParameterList")
    external fun nMakeVk(
        width: Int,
        height: Int,
        sampleCount: Int,
        mipmapped: Boolean,
        imageCreateFlags: Int,
        format: Int,
        imageTiling: Int,
        imageUsageFlags: Int,
        sharingMode: Int,
        aspectMask: Int,
        currentLayout: Int,
        queueFamilyIndex: Int,
        image: Long,
        allocMemory: Long,
        allocOffset: Long,
        allocSize: Long,
        allocFlags: Int,
    ): Long

    external fun nMakeMetal(
        width: Int,
        height: Int,
        texture: Long,
    ): Long

    external fun nRelease(ptr: Long)

    external fun nIsValid(ptr: Long): Boolean
}

File: rust/skialin-core/src/graphite_backend_texture.rs
use crate::sys;

/// Wraps a skgpu::graphite::BackendTexture: a caller-owned GPU texture (VkImage, or an
/// ID3D12Resource imported through Dawn) for rendering into via
/// `Surface::wrap_graphite_backend_texture`. The second field keeps alive whatever the native
/// BackendTexture borrows without retaining (Dawn's imported texture); fields drop after `Drop::drop`
/// runs, so it outlives the BackendTexture itself.
pub struct GraphiteBackendTexture(pub(crate) *mut sys::skgpu::graphite::BackendTexture, #[allow(dead_code)] pub(crate) Option<Box<dyn std::any::Any>>);

impl GraphiteBackendTexture {
    #[allow(clippy::too_many_arguments)]
    pub fn new_vk(
        width: i32,
        height: i32,
        sample_count: i32,
        mipmapped: bool,
        image_create_flags: u32,
        format: sys::VkFormat,
        image_tiling: sys::VkImageTiling,
        image_usage_flags: sys::VkImageUsageFlags,
        sharing_mode: sys::VkSharingMode,
        aspect_mask: sys::VkImageAspectFlags,
        current_layout: sys::VkImageLayout,
        queue_family_index: u32,
        image: sys::VkImage,
        alloc_memory: sys::VkDeviceMemory,
        alloc_offset: sys::VkDeviceSize,
        alloc_size: sys::VkDeviceSize,
        alloc_flags: u32,
    ) -> Self {
        let ptr = unsafe {
            sys::skialin_bridge_GraphiteBackendTexture_MakeVk(
                width,
                height,
                sample_count,
                mipmapped,
                image_create_flags,
                format,
                image_tiling,
                image_usage_flags,
                sharing_mode,
                aspect_mask,
                current_layout,
                queue_family_index,
                image,
                alloc_memory,
                alloc_offset,
                alloc_size,
                alloc_flags,
            )
        };
        GraphiteBackendTexture(ptr, None)
    }

    pub fn is_valid(&self) -> bool {
        unsafe { sys::skialin_bridge_GraphiteBackendTexture_isValid(self.0) }
    }
}

impl Drop for GraphiteBackendTexture {
    fn drop(&mut self) {
        unsafe { sys::skialin_bridge_GraphiteBackendTexture_delete(self.0) };
    }
}
